src/web/demo.rkt

```diff
@@ -213,45 +213,89 @@
-          (define test (parse-test formula))
-          (eprintf "Sampling Job ~a started:\n  sample ~a...\n" job-id (syntax->datum formula))
-          (define result (run-herbie 'sample test #:seed seed* #:profile? #f #:timeline-disabled? #t))
-          (hash-set! *completed-jobs* job-id result)
-          (eprintf " complete\n")
-          (hash-remove! *jobs* job-id)
-          (semaphore-post sema)])
+         [job-info (run-job job-info)])
        (loop seed)))))
 
+(struct run-herbie-command 
+ (command formula seed pcontext profile? timeline-disabled?))
+
+(define (wrapper-run-herbie cmd job-id after-job-work)
+  (print-job-message (run-herbie-command-command cmd) job-id (syntax->datum (run-herbie-command-formula cmd)))
+  (define result (run-herbie 
+   (run-herbie-command-command cmd)
+   (parse-test (run-herbie-command-formula cmd))
+   #:seed (run-herbie-command-seed cmd)
+   #:pcontext (run-herbie-command-pcontext cmd)
+   #:profile? (run-herbie-command-profile? cmd)
+   #:timeline-disabled? (run-herbie-command-timeline-disabled? cmd)))
+  (hash-set! *completed-jobs* job-id result)
+  (after-job-work result job-id (run-herbie-command-seed cmd)))
+
+(define (after-improve result job-id seed)
+ (define path (format "~a.~a" job-id *herbie-commit*))
+ (when (*demo-output*)
+    ;; Output results
+    (make-directory (build-path (*demo-output*) path))
+    (for ([page (all-pages result)])
+      (call-with-output-file (build-path (*demo-output*) path page)
+        (λ (out) 
+          (with-handlers ([exn:fail? (page-error-handler result page out)])
+            (make-page page out result (*demo-output*) #f)))))
+    (update-report result path seed
+                    (build-path (*demo-output*) "results.json")
+                    (build-path (*demo-output*) "index.html"))))
+
+; A place holder helper function that should be used when no work needs to be
+; done on the finished result
+(define (default-after result job-id seed) empty)
+
+(define (run-job job-info)
+ (define sema (work-sema job-info))
+ (define job-id (work-id job-info))
+ (define path (format "~a.~a" job-id *herbie-commit*))
+ (cond ;; Check caches if job as already been completed
+  [(hash-has-key? *completed-jobs* job-id)
+  (semaphore-post sema)]
+  [(and (*demo-output*) (directory-exists? (build-path (*demo-output*) path)))
+  (semaphore-post sema)]
+  [else (match (work-info job-info)
+    [(list 'improve formula)
+      (wrapper-run-herbie 
+       (run-herbie-command 'improve formula (get-seed) #f #f #f) 
+       job-id after-improve)]
+    [(list 'sample formula seed*)
+      (wrapper-run-herbie 
+       (run-herbie-command 'sample formula seed* #f #f #t) 
+       job-id default-after)])])
+ (eprintf "Job ~a complete\n" job-id)
+ (hash-remove! *jobs* job-id)
+ (semaphore-post sema))
+
+; Handles semaphore and async part of a job
+(struct work (id info sema))
+
+; Encapsulates semaphores and async part of jobs.
+(define (run-work sync-job job-id job)
```

Comment:
Why does this method require you to pass in the `job-id`? Why not pass in the job & have it compute the job id for you? Perhaps it could return the `work` object so the caller could retrieve the job id if it needed it.

Reply:
Style nits:
- `sync-job` is a boolean and should be named with a question mark, like `sync-job?`
- Also it seems safe to pick a default and make it a keyword argument, as in `#:sync? [sync-job? #t]`